working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+parameters cannot have those modifiers). This omission is not seen as a problem
+in this proposal: It is always possible to use a normal constructor declaration
+and normal instance variable declarations, and it is probably a useful property
+that the primary constructor uses a formal parameter syntax which is completely
+like that of any other formal parameter list. Just use a normal declaration. Use
+an initializing formal in a primary constructor to initialize it from the
+primary constructor, if needed.
+
+An `external` instance variable amounts to an `external` getter and an
+`external` setter. Such "variables" cannot be initialized by an
+initializing formal anyway, so they do not fit into the treatment implied
+by a primary constructor. Just use a normal declaration.
+
+```dart
+class ModifierClass {
+  late int x;
+  external double d;
+  ModifierClass(this.x);
+}
+
+class ModifierClass(this.x) {
+  late int x;
+  external double d;
+}
+```
+
+`ModifierClass` as written does not make sense (`x` does not have to be
+`late`), but there could be other constructors that do not initialize `x`.
+
+Super parameters can be declared in the same way as in non-primary
+constructors:
+
+```dart
+class A {
+  final int a;
+  A(this.a);
+}
+
+class B extends A {
+  B(super.a);
+}
+
+class A(final int a);
+class B(super.a) extends A;
+```
+
+Next, the constructor can be named, and it can be constant:
+
+```dart
+class Point {
+  final int x;
+  final int y;
+  const Point._(this.x, this.y);
+}
+
+class const Point._(final int x, final int y);
+```
+
+Note that the class header contains syntax that resembles the constructor
+declaration, which may be helpful when reading the code.
+
+The modifier `const` could have been placed on the class (`const class`)
+rather than on the class name. This proposal puts it on the class name
+because the notion of a "constant class" conflicts with with actual
+semantics: It is the constructor which is constant because it is able to be
+invoked during constant expression evaluation; it can also be invoked at
+run time, and there could be other (non-constant) constructors. This means
+that it is at least potentially confusing to say that it is a "constant
+class", but it is consistent with the rest of the language to say that this
+particular primary constructor is a "constant constructor". Hence `class
+const Name` rather than `const class Name`.
+
+The modifier `final` on a parameter in a primary constructor has no meaning
+for the parameter itself, because there is no scope where the parameter can
+be accessed. Hence, this modifier is used to specify that the instance
+variables declared by this primary constructor are `final`.
+
+In the case where the constructor is constant, and in the case where the
+declaration is an `inline` class or an `enum` declaration, the modifier
+`final` on every instance variable is required. Hence, it can be omitted:
```

Comment:
I think we should resolve that issue in general. When we know what we _want_ to do with formal parameters using `var`, we may introduce that treatment in general, and primary constructors is just one case among others.

(I do _not_ think it's an attractive idea to use `var` to play a special role for the formal parameters of a primary constructor just because "they are about variables as well as parameters, so we can use `var` to mean [something]". Primary constructors are currently using formal parameters that are syntactically and semantically the same as other formal parameters, and I think it would be nice if we can preserve that property.)

The current state is that a formal parameter declaration in an instance method whose first token is `var` yields a parameter type which is determined by override inference, falling back to `dynamic` if there's no overriding relation. In functions that aren't instance methods, it means `dynamic`. (This treatment is not fully specified, as far as I can see, so I'm just reporting the behaviors that I can see via a few experiments).

One obvious way to go is to say that a formal parameter declaration whose first token is `var` is a compile-time error. We could lint/warn against that now, and then make it an error later on, and in that case it seems perfectly fine to make it an error for primary constructor parameters now.

Note that we have at least many hundreds of examples of `var` used as a "formal parameter type" in existing code, which means that we will need to consider some actual breakage.

Reply:
I'll assume that we handle this question separately, dealing with _all_ parameters using `var` at the same time, or just the ones in primary constructors, but even that can be handled in a discussion issue separately from the first version of the feature spec proposal.

Reply:
So I'll resolve the thread.